honeybee/aperture.py

```diff
@@ -0,0 +1,506 @@
+# coding: utf-8
+"""Honeybee Aperture."""
+from .properties import ApertureProperties
+from .aperturetype import aperture_types, Window
+from .boundarycondition import Outdoors, Surface
+from .shade import Shade
+from .typing import valid_string
+import honeybee.writer as writer
+
+from ladybug_geometry.geometry3d.pointvector import Point3D, Vector3D
+from ladybug_geometry.geometry3d.face import Face3D
+
+import math
+
+
+class Aperture(object):
+    """A single planar aperture face.
+
+    Properties:
+        name
+        name_original
+        geometry
+        vertices
+        upper_left_vertices
+        triangulated_mesh3d
+        normal
+        center
+        area
+        perimeter
+        boundary_condition
+        parent
+        has_parent
+    """
+    TYPES = aperture_types
+    __slots__ = ('_name', '_name_original', '_geometry', '_parent',
+                 '_type', '_boundary_condition', '_properties', '_writer')
+
+    def __init__(self, name, geometry, type=None, boundary_condition=None):
+        """A single planar aperture in a face.
+
+        Args:
+            name: Aperture name. Must be < 100 characters.
+            geometry: A ladybug-geometry Face3D.
+            type: Aperture type. Default: Window
+            boundary_condition: Boundary condition object (Outdoors, Surface).
+                Default: Outdoors.
+        """
+        # process the name
+        self._name = valid_string(name, 'honeybee aperture name')
+        self._name_original = name
+
+        # process the geometry
+        assert isinstance(geometry, Face3D), \
+            'Expected ladybug_geometry Face3D. Got {}'.format(type(geometry))
+        self._geometry = geometry
+        self._parent = None  # _parent will be set when the Face is added to a Face
+
+        # process the type and boundary condition
+        self.type = type or Window()
+        self.boundary_condition = boundary_condition or Outdoors()
```

Comment:
Ah, I didn't think of that.  Ok, I will do this in all places where we have types of face/aperture and boundary conditions that aren't "Surface".

Reply:
Actually, wait. The one thing to keep in mind here is that, if someone decides to edit the `sun_exposure` or `view_factor` of the default outdoor boundary condition assigned to the Aperture, it will set all of the defaulted outdoor boundary conditions across the plugin to not have sun exposure.  Is this what we want or should we be making new instances of `Outdoors` for each aperture like I have here?  Or should we just have the boundary conditions locked whenever they get assigned to an Aperture, Face or Door.  That way, people can always make a custom boundary condition and assign it to a Face but the can't edit the assigned boundary condition directly and accidentally edit all of the default boundary conditions.

Reply:
I think the locking actually won't help us too much in the plugin because, when someone goes to duplicate a Face, the boundary condition will have to be duplicated as well unless we want to add an extra property to boundary condition in to keep track of whether the boundary condition is default.  So, after the first duplication, we will end up with tons of instances of the boundary condition anyway.
Perhaps the best solution is just to remove the `sun_exposure` and `wind_exposure` setters for the boundary condition (essentially making them immutable and thereby tying the exposure to the type of boundary condition). Along these lines, we can remove the 'view_factor` property since it seems dumb to compromise the whole usage of the boundary condition object across the library just to satisfy the one researcher who wants to set this property.  Honestly, that one researcher can just write their own measure.

What do you think?

Reply:
I think making sun_exposure and wind_exposure and view factor read_only is a good idea. I don't think we should remove them though.

Using locked might be an overkill in this case.

One can still make a custom boundary condition from the base class if needed.

Reply:
Same comments as above. I think overwriting the setters and making them read only makes the most sense and users can create their own custom boundary condition if needed for cases like view factor.

Reply:
Sounds good and I agree.  I definitely wasn't thinking of removing the `sun_exposure` and `wind_exposure` properties (I just wanted to remove the setters such that they become read-only).  I originally didn't think about exposing them on `__init__` because they aren't exposed there now but, now that you mention it, that seems like a really good idea.  So people can always create the custom boundary condition and assign it to a Face if they really want to.